maven/poms.bzl

```diff
@@ -78,7 +108,68 @@ def _extract_dependencies(parent_node):
 # The parameter should be an xml node containing the tag <dependencyManagement>
 def _extract_dependency_management(project_node):
     node = xml.find_first(project_node, labels.DEPENDENCY_MANAGEMENT)
-    return poms.extract_dependencies(node) if bool(node) else []
+    return _extract_dependencies(node) if bool(node) else []
+
+def _merge_dependency(fallback, dependency):
+    if not bool(fallback):
+        return dependency
+
+    # If this were a node tree, we could re-use the leaf element merging, but it's a struct. :(
+    # groupid and artifactid are always present, so don't bother merging them.
+    return _dependency(
+        group_id = dependency.group_id,
+        artifact_id = dependency.artifact_id,
+        version = dependency.version if bool(dependency.version) else fallback.version,
+        type = dependency.type if bool(dependency.type) else fallback.type,
+        optional = dependency.optional if bool(dependency.optional) else fallback.optional,
+        scope = dependency.scope if bool(dependency.scope) else fallback.scope,
+        classifier = dependency.classifier if bool(dependency.classifier) else fallback.classifier,
+        system_path = dependency.system_path if bool(dependency.system_path) else fallback.system_path,
+    )
+
+def _get_variable(string):
+    start = string.find("${")
+    end = string.find("}")
+    return string[start + 2:end] if start >= 0 and end >= 0 and start < end - 3 else None
+
+def _substitute_variable_if_present(string, properties):
+    if not bool(string):
+        return None
+    variable_label = _get_variable(string)
+    if bool(variable_label):
+        substitution = properties.get(variable_label, None)
+        if bool(substitution):
+            string = string.replace("${%s}" % variable_label, substitution)
+    return string
+
+def _apply_property(dependency, properties):
+    # For now just check <version> and <groupId>
```

Comment:
How long is "For now"? Do you anticipate more subs?

Reply:
Shouldn't `artifact_id` get this treatment too? Asking since it's one of the implicit properties in a for loop below.

Reply:
No, it is one of the only things that is never set in a variable, because it's basically useless to set.  You get some payoff saying "take X artifact from my group" but you never have a good reason to say "take artifact named like me from x other group". In 20+ years of maven, I've never seen artifact ID specified with a substitution. And I couldn't find one with a web search or corpus search on sonatype.